### PR TITLE
Trapezoid Generator Function 

### PR DESCRIPTION
**Pull request by Skandis**

### CODE REVIEW

**Review by auscompgeek — CHANGES_REQUESTED** [review]

File: utilities/pure_pursuit.py [hunk 1]
```diff
@@ -21,6 +21,8 @@ def __init__(self, look_ahead):
         self.look_ahead = look_ahead
         self.completed_path = False
         self.distance_traveled = 0
+        self.trapezoid = False
+        self.max_speed = 3  # <-- change to actuall max speed when avaliable
```
Comment (auscompgeek) [inline utilities/pure_pursuit.py]: Don't hardcode this. Make this an input.

File: utilities/pure_pursuit.py [hunk 2]
```diff
@@ -73,9 +75,9 @@ def find_intersections(self, waypoint_start, waypoint_end, robot_position):
             print("No intersection found")
 
     def build_path(self, waypoints):
```
Comment (auscompgeek) [inline utilities/pure_pursuit.py]: You cannot have code before a docstring. That makes it no longer a docstring.

File: utilities/pure_pursuit.py [hunk 3]
```diff
@@ -73,9 +75,9 @@ def find_intersections(self, waypoint_start, waypoint_end, robot_position):
...
+        current_waypoint = 0
         """
         Take in a list of waypoints used to build a path.
```
Comment (auscompgeek) [inline utilities/pure_pursuit.py]: Why have you removed this line?

File: utilities/pure_pursuit.py [hunk 4]
```diff
@@ -96,6 +98,55 @@ def build_path(self, waypoints):
...
+            angle = math.atan2(end_point[1], end_point[0])
+            sin_x = math.sin(angle)
+            cos_y = math.cos(angle)
+
+            if self.trapezoid == True:
+                x, y, speed = waypoints[current_waypoint]
+                waypoint_distance += math.hypot(
+                    x - previous_waypoint[0], y - previous_waypoint[1]
+                )
+                accel, decel, speed = self.generate_trapezoidal_function(
```
Comment (auscompgeek) [inline utilities/pure_pursuit.py]: I don't think you should be hardcoding this either.

File: utilities/pure_pursuit.py [hunk 5]
```diff
@@ -96,6 +98,55 @@ def build_path(self, waypoints):
             self.waypoints.append((x, y, speed, waypoint_distance))
         self.current_waypoint_number = 0
         print(self.waypoints)
+        if current_waypoint <= len(self.waypoints):
+            end_point = (
+                self.waypoints[current_waypoint + 1] - self.waypoints[current_waypoint]
+            )
+            angle = math.atan2(end_point[1], end_point[0])
+            sin_x = math.sin(angle)
+            cos_y = math.cos(angle)
```
Comment (auscompgeek) [inline utilities/pure_pursuit.py]: Don't compare to a bool.

File: utilities/pure_pursuit.py [hunk 6]
```diff
@@ -96,6 +98,55 @@ def build_path(self, waypoints):
...
+                    (start_cruise, self.max_speed, waypoint_distance),
+                )
+                self.waypoints.insert(
+                    current_waypoint + 2,
+                    (start_decel, self.max_speed, waypoint_distance),
+                )
+                current_waypoint += 3
+                print("new waypoints")
+                print("")
+                print(waypoints)
```
Comment (auscompgeek) [inline utilities/pure_pursuit.py]: Don't compare to a bool.

If `self.trapezoid` is always a bool, make this an else.

File: utilities/pure_pursuit.py [hunk 7]
```diff
@@ -173,3 +224,29 @@ def find_velocity(self, robot_position):
         vx, vy = direction * speed
         heading = 0
         return vx, vy, heading
```
Comment (auscompgeek) [inline utilities/pure_pursuit.py]: Does this actually generate a function?

File: utilities/pure_pursuit.py [hunk 8]
```diff
@@ -225,8 +229,8 @@ def find_velocity(self, robot_position):
...
 
-    def generate_trapezoidal_function(self, v, u, a, d):
-        """Generate a trapezoidal function for acceleration.
+    def generate_traezoidal_distances(self, v, u, a, d):
+        """Generate how far you have to travle to acceleration and deceleration for speed control.
```
Comment (auscompgeek) [inline utilities/pure_pursuit.py]: speeel and grandma check

**Comment by james-ward** [comment]
Have a look at my fork on the trapezoid branch:
https://github.com/james-ward/pydeepspace/tree/trapezoid

The tests are broken, but I think this is a nicer way to do it.

**Pull request closed without merging**